bellows/zigbee/application.py

```diff
@@ -421,35 +511,46 @@ async def write_network_info(
         assert status == t.EmberStatus.SUCCESS
 
         # Write APS link keys
-        for key in network_info.key_table:
-            ember_key = util.zigpy_key_to_ezsp_key(key, ezsp)
+        for index, key in enumerate(network_info.key_table):
+            if ezsp.ezsp_version < 13:
+                # XXX: is there no way to set the outgoing frame counter or seq?
+                (status,) = await ezsp.addOrUpdateKeyTableEntry(
+                    key.partner_ieee, True, key.key
+                )
+            else:
+                (status,) = await ezsp.importLinkKey(index, key.partner_ieee, key.key)
 
-            # XXX: is there no way to set the outgoing frame counter or seq?
-            (status,) = await ezsp.addOrUpdateKeyTableEntry(
-                ember_key.partnerEUI64, True, ember_key.key
-            )
             if status != t.EmberStatus.SUCCESS:
                 LOGGER.warning("Couldn't add %s key: %s", key, status)
 
-        if ezsp.ezsp_version > 4:
-            # Set NWK frame counter
-            (status,) = await ezsp.setValue(
-                ezsp.types.EzspValueId.VALUE_NWK_FRAME_COUNTER,
-                t.uint32_t(network_info.network_key.tx_counter).serialize(),
-            )
-            assert status == t.EmberStatus.SUCCESS
-
-            # Set APS frame counter
-            (status,) = await ezsp.setValue(
-                ezsp.types.EzspValueId.VALUE_APS_FRAME_COUNTER,
-                t.uint32_t(network_info.tc_link_key.tx_counter).serialize(),
-            )
-            assert status == t.EmberStatus.SUCCESS
+        # Write the child table
+        if ezsp.ezsp_version >= 9:
+            index = 0
+
+            for child_eui64 in self.state.network_info.children:
+                if child_eui64 not in self.state.network_info.nwk_addresses:
+                    continue
+
+                status = await ezsp.setChildData(
+                    index,
+                    ezsp.types.ChildData(
+                        eui64=child_eui64,
+                        # XXX: why does EmberZNet also have `MOBILE_END_DEVICE`?
```

Comment:
The SDK also defines `MOBILE_END_DEVICE` but I can't figure out what this refers to.

@MattWestb have you heard of this before?

Reply:
Looks like a proprietary extension.

> Mobile end devices (EmberNodeType EMBER_MOBILE_END_DEVICE) is a sleepy end device with enhanced capabilities that enable it to change its physical location and quickly switch to a new parent router. This device type is an Ember modification to the basic ZigBee sleepy end device to provide added capabilities and management of mobile devices.


> Mobile End Devices - For devices that change parents (or move out of parent range) often, they can join as EMBER_MOBILE_END_DEVICE (selectable in AppBuilder) and our stack will store the child entries in RAM with a more aggressive timeout (EMBER_MOBILE_NODE_POLL_TIMEOUT) rather than writing to SimEEPROM memory each time the child comes and goes.

I think using `SLEEPY_END_DEVICE` is better, as in zigpy we don't keep track of the child type.

Reply:
```
There are three types of sensor nodes: line powered sensors, sleepy
//  sensors, and mobile sensors. Line powered sensors route and are always
//  awake. They can also serve as parents to sleepy or mobile sensors.
//  Sleepy sensors sleep whenever they can to conserve power but they do
//  not route and rely on their parent when communicating with the network.
//  Mobile sensors are like sleepy sensors but are meant to move. mobile
//  sensors will wait a smaller amount of time (5 seconds) to switch parents
//  than sleepy sensors will (60 seconds) if their parent does not respond to
//  their polls.
```
I found this in a header for one of their samples